src/fastaq_handler.cpp

```diff
@@ -1,176 +1,83 @@
+        throw "Truncated quality string detected";
+    } else if (read_status == -3) {
+        throw "Error reading " + this->filepath;
     }
 
-    while (getline(instream, line).good()) {
-        if (!line.empty() and line[0] == '+') {
-            // skip this line and the qual score line
-            getline(instream, line);
-            // cout << "qual score line ." << line << "." << endl;
-        } else if (line.empty() or line[0] == '>' or line[0] == '@') {
-            return;
-        }
-    }
+    ++this->num_reads_parsed;
+    this->name = this->inbuf->name.s;
+    this->read = this->inbuf->seq.s;
 }
 
-void print(std::ifstream& infile)
+void FastaqHandler::get_nth_read(const uint32_t& idx)
 {
-    char file;
-    std::vector<char> read;
-    uint i = 0;
-
-    // Read infile to vector
-    while (!infile.eof()) {
-        infile >> file;
-        read.push_back(file);
+    // edge case where no reads have been loaded yet
+    if (this->num_reads_parsed == 0) {
+        this->get_next();
     }
-
-    // Print read vector
-    for (i = 0; i < read.size(); i++) {
-        std::cout << read[i];
-    }
-}
-
-void print(std::istream& infile)
-{
-    char file;
-    std::vector<char> read;
-    int i = 0;
-
-    // Read infile to vector
-    while (!infile.eof()) {
-        infile >> file;
-        read.push_back(file);
-    }
-
-    // Print read vector
-    for (auto i = 0; i < read.size(); i++) {
-        std::cout << read[i];
-    }
-}
-
-void FastaqHandler::get_id(const uint32_t& id)
-{
-    const uint32_t one_based_id = id + 1;
-    if (one_based_id < num_reads_parsed) {
-        BOOST_LOG_TRIVIAL(warning)
-            << "restart buffer as have id " << num_reads_parsed << " and want id "
-            << one_based_id << " (" << id << ") with 0-based indexing.";
+    const uint32_t one_based_idx = idx + 1;
+    if (one_based_idx < this->num_reads_parsed) {
         num_reads_parsed = 0;
         name.clear();
         read.clear();
-        line.clear();
-        assert(name.empty());
-        assert(read.empty());
-        assert(line.empty());
-
-        instream.ignore(std::numeric_limits<std::streamsize>::max());
-        fastaq_file.seekg(0, fastaq_file.beg);
-        instream.clear();
-        inbuf.pop();
-        if (gzipped) {
-            inbuf.pop();
-            inbuf.push(boost::iostreams::gzip_decompressor());
-        }
-        inbuf.push(fastaq_file);
-    }
-
-    while (id > 1 and num_reads_parsed < id) {
-        skip_next();
-        if (eof()) {
-            break;
-        }
+        gzrewind(this->fastaq_file);
+        kseq_rewind(this->inbuf);
     }
 
-    while (num_reads_parsed <= id) {
-        get_next();
-        if (eof()) {
-            break;
+    while (this->num_reads_parsed < one_based_idx) {
+        try {
+            this->get_next();
+        } catch (std::out_of_range& err) {
+            throw std::out_of_range("Requested a read past the end of the file.");
         }
```

Comment:
Very minor: I think there is no need to enclose this `get_next()` in a `try` `catch` block as it is catching an exception, rebuilding and rethrowing it